### ISSUE TITLE
Kanidm (2024)

### ISSUE THREAD

**Issue by Firstyear** [openSUSE/mentoring#196]
**Project Title**: Kanidm

**Description**: Kanidm is an identity management project written in Rust. IDM services are a network store of users and groups for systems, websites and more, providing these with authentication and authorisation functionality. It aims to respect all peoples' identities and to store their data safely and correctly. IDM is a diverse area which values contributions from many people, from a variety of backgrounds. 

**Assessment**: Our intent is to mentor and help you grow as a contributor - and we want to help you succeed! As this is our focus, we have different assessment criteria to many other projects.

Has the student:
* participated actively in the project?
* written and submitted personal blog/journal or similar about the experience and what they have learnt and attempted?
* sufficiently attempted one or any of the offered deliverables? (Note - there is no requirement to *complete* the deliverable. Just evidence of a solid attempt)

**Deliverable**: As IDM spans many areas, we offer multiple possible focus areas. You should select one of these areas.

There is no requirement to complete *all* of these - we want you to attempt to work in an area of interest to you, and if you wish to change focus during the project, we can negotiate altering or changing which deliverable is your focus. 

Possible deliverable areas (choose one):

* Documentation and Accessibility - Improvements to project documentation, --help text, design of the user interface and command line tools, and extension of the CLI with new functionality.
* Database Performance - creation of a statistics collection and analysis process that can assist a query optimiser in improving query performance. 
* Rust FFI for RADIUS - rewrite the current RLM python module into Rust, with C-ffi bindings. 
* WebUI - Improve the web interface to allow users an admins to perform more tasks
* Event Auditing - Log specific event's to a dedicated auditing log for analysis
* Your choice - there are many ideas in our issue tracker which you can choose from. Alternately, suggest your own! https://github.com/kanidm/kanidm/issues

**Mentor**: William Brown (wbrown@suse.de - UTC+10:00).

**Skills**: Basic Programming. Nice to have - some knowledge of databases and Rust (but not required). 

**Skill Level**: Beginner to Advanced. Each topic can be scaled up or down to your level, and areas of interest.

**Project Size:** Medium to Large - Again it can be scaled on your interest and direction. 

**Get started**: 

The first step is to attempt to setup an instance of Kanidm to understand what the project does and what it offers. The following documentation will help you further with this. 

[1] https://github.com/kanidm/kanidm/blob/master/README.md
[2] https://github.com/kanidm/kanidm/blob/master/CODE_OF_CONDUCT.md
[3] https://kanidm.github.io/kanidm/master/ 
[4] https://github.com/kanidm/kanidm/blob/master/book/src/developers/readme.md

**Comment by harshalmittal4** [comment]
Hello @Firstyear , @ddemaio , I am exploring this project and found it interesting. I wish to contribute to this idea as part of GSoC, could you please tell if it's for GSoC?
Can you also please help me with finding some specific issues that I must start with. I found [this](https://github.com/kanidm/kanidm/issues/1198) issue on checking the good-first-issues and wished to get more context about which profile page need to be updated.

Thanks!

**Comment by ddemaio** [comment]
Hi @harshalmittal4 These are listed here for GSoC. @Firstyear can provide input. I do the admin for GSoC. Great to see your interest  in the project.

**Comment by harshalmittal4** [comment]
Thank you @ddemaio, hey @Firstyear looking forward to your advice :)

**Comment by Firstyear** [comment]
@harshalmittal4 I'd rather you try to understand what the project does and how it works first before trying to solve issues, so please have a look through the project spec here again and setup an instance first.

**Comment by uncomfyhalomacro** [comment]
Hello @Firstyear. Thanks for pointing me to this project. I have to explore around the project and issue tracker that fits my skill level. I am pinging @rv178 since this is a friend of mine who has more potential than me and has experience in Rust. I want him to join GSoC too in case he is interested 🥰

Anyway, looking forward to contribute but first I am going to read and understand the idea of this project. :heart:

**Comment by harshalmittal4** [comment]
Thanks @Firstyear. I will set it up first, was going through the kanidm docs.

**Comment by Sam071100** [comment]
Hello @Firstyear, @ddemaio, I was going through this project and found it interesting. I wish to contribute to this idea as part of GSoC. I have good knowledge in Rust programming language as I have previously contributed to the Summer of Bitcoin'21. As of now, I am trying to understand what the project does. I am personally greatly interested in this project and keen to work under your guidance.

**Comment by OfficialAkashKumar** [comment]
Hello @Firstyear , I came across this project in Gsoc 2023 under openSUSE and it seems quite interesting to me. I have intermediate knowledge in rust and mysql expertise in solving algorithmic problems.

I would like to work on Rust FFI for RADIUS and Identity verification.

**Comment by czzrr** [comment]
Hey @Firstyear and @ddemaio! My name is Christopher. I'm studying Computer Science & Engineering in Denmark. :)
Very interesting project! My plan is to start contributing and look into the project ideas. Looking forward to collaborate!

**Comment by Firstyear** [comment]
> I would _really_ enjoy contributing, specifically in the Identity Verification area, as I'm especially interested in rust and cybersecurity! 

Sure, please get in touch on the issue or the community chat channel, it's listed in the readme.

**Comment by Seba-T** [comment]
I tried to fix #1359, I hope it follow all the guidelines! (Btw I was unsure whether merging with upstream before making the PR or leaving it as it was, at the end I merged, I hope it's okay)

**Comment by Firstyear** [comment]
I've commented on the PR

**Comment by MinhPhan8803** [comment]
Hello @ddemaio @Firstyear, my name is Minh and I came here from GSOC. I have working knowledge of Rust and SQL, having used them in various projects. I found this project very interesting and would love to contribute to it!

**Comment by PythonHacker24** [comment]
Hey @Firstyear @ddemaio, I have been going through the **OpenSUSE Project** for a while now to get into the **GSOC 2024**. I found this project to be the one that I would like to work on. Since this project is related to authentication which coincides with my CyberSecurity Domain (which I have been self-studying for almost a year now) as well as it has its codebase in RUST that I am currently working on. I have a working understanding of databases as I manage the system design for my organisation, so working on authentication systems would be of interest.

**Comment by Firstyear** [comment]
I haven't yet committed to GSoC 2024, so I can't really comment on this right now. This was about the 2023 program. Sorry.

**Comment by PythonHacker24** [comment]
No worries will see you in GSOC 2024. I think it's nice to start early.

**Comment by Firstyear** [comment]
I'm not even sure I'm going to mentor in 2024, so I think you're *way* too early here 😅 There are other ways to contribute to projects that aren't just gsoc I hope you know.

**Comment by PythonHacker24** [comment]
Yes, I am aware of the contributions that I can make, which aren't GSOC related. I am actually new to the Open Source Community and will look forward to making good contributions. I am going through the docs as well as the source code. I also went through your conference videos on Kanidm on the OpenSUSE channel. It was awesome learning from you about the past authentication solutions and the difference that Kanidm is making by solving the flaws they had. (big fan of the SSH authentication mechanism with QR code, I never thought something like that!)

**Comment by Firstyear** [comment]
Have a look at our developer guide ( https://kanidm.github.io/kanidm/stable/DEVELOPER_README.html ) and also the matrix room ( https://gitter.im/kanidm/community  )

**Comment by PythonHacker24** [comment]
Sure, will have a look at it.

**Comment by AnsahMohammad** [comment]
Hey @Firstyear,
I'm Ansah, a CS undergrad keen on the tech side of projects. Discovered Kanidm for GSoC 2024, and it's caught my eye. 

I've got a basic handle on Rust and solid DB experience (SQL, NoSQL).I have Previously been a  QAMP Mentorship, Qhack2024 and has contributed to Appflowy. Your guidance would mean a lot as I dive in!.

**Comment by Firstyear** [comment]
Hi there,

Please have a look at and follow the guide in the issue description to get started, the project readme and developer guides especially.

**Comment by ddemaio** [comment]
Please reopen @Firstyear if you want to mentor this in GSoC 2025.

**Comment by Firstyear** [comment]
@ddemaio Thanks mate, I don't want to mentor in 2025, but maybe the year after.
